api/oc_push.c

```diff
@@ -0,0 +1,2555 @@
+ * Resource
+ */
+OC_MEMB(ns_instance_memb, oc_ns_t, 1);
+
+/*
+ * @brief	`ns_col_list` keeps real data of all Notification Selector Resources
+ * 			(it includes all Resources of all Devices)
+ *
+ * 			each list member is instance of `oc_ns_t`
+ */
+OC_LIST(ns_list);
+
+/*
+ * @brief	memory block definition for storing new Receiver object array of Push
+ * Receiver Resource
+ */
+OC_MEMB(recvs_instance_memb, oc_recvs_t, 1);
+
+/*
+ * @brief	memory block definition for storing new Receiver object of Receiver
+ * object array
+ */
+OC_MEMB(recv_instance_memb, oc_recv_t, 1);
+
+/*
+ * @brief	`recvs_list` keeps real data of all Receiver object in Push Receiver
+ * Resource (it includes all Receiver objects of Resource of all Devices)
+ *
+ * 			each list member is instance of `oc_recvs_t`
+ */
+OC_LIST(recvs_list);
+
+/*
+ * @brief	memory block definition for storing properties representation of
+ * pushed resource
+ */
+OC_MEMB(rep_instance_memb, oc_rep_t, 1);
+
+/*
+ * @brief	memory block definition for storing pushed resource representation
+ * list
+ */
+OC_MEMB(pushd_rsc_rep_instance_memb, oc_pushd_rsc_rep_t, 1);
+
+/*
+ * @brief	`pushed_rsc_list` keeps Resource representation of Pushed Resources
+ */
+OC_LIST(pushd_rsc_rep_list);
+
+/*
+ * @brief	process which handles push notification
+ */
+OC_PROCESS(oc_push_process, "Push Notification handler");
+
+const char *pp_state_strs[] = {
+  "waitingforprovisioning",       /*OC_PP_WFP*/
+  "waitingforupdate",             /*OC_PP_WFU*/
+  "waitingforresponse",           /*OC_PP_WFR*/
+  "waitingforupdatemitigation",   /*OC_PP_WFUM*/
+  "waitingforresponsemitigation", /*OC_PP_WFRM*/
+  "error",                        /*OC_PP_ERR*/
+  "timeout"                       /*OC_PP_TOUT*/
+};
+
+/*
+ * mandatory property of oic.r.pushporxy, oic.r.pushreceivers
+ */
+enum {
+  PP_PUSHTARGET = 0x01,
+  PP_SOURCERT = 0x02,
+  PR_RECEIVERS = 0x01,
+  PR_RECEIVERURI = 0x02,
+  PR_RTS = 0x04
+};
+
+/*
+ * if this callback function is provided by user, it will called whenever new
+ * push is arrived...
+ */
+static void (*oc_push_arrived)(oc_pushd_rsc_rep_t *) = NULL;
+
+#define pp_statestr(i) (pp_state_strs[(i)])
+
+/*
+ * @brief update Push Proxy state from state to new_state
+ *
+ * @param state		oc_string_t
+ * @param new_state	char *
+ */
+#define pp_update_state(state, new_state)                                      \
+  (oc_set_string(&(state), (new_state), strlen((new_state))))
+
+/*
+ * @brief initialize oc_string_t object
+ */
+#define oc_init_string(str)                                                    \
+  {                                                                            \
+    (str).size = 0;                                                            \
+    (str).ptr = NULL;                                                          \
+    (str).next = NULL;                                                         \
+  }
+
+OC_API
```

Comment:
Does this work? I always thought the annotation needed to be in the header, but this way would be nicer actually, less clutter in the header files. I'm going to make a note and see if it works.

Reply:
I think so. 
I found many example using visibility in function definition

Reply:
Indeed, seems to work on Unix (I've created a `.so` library and then invoked `nm -D` on it and the symbol was visible). Good to know.

However, I think this will be a problem on Windows with MSVC. If I understand it correctly, the exported function must be annocated with `dllExport` in the library itself and then with `dllImport` when the library is used in a binary. And this can only be done in the header.

Reply:
ok, I'll leave it in header only

Reply:
done